File: src/lib/routing/routeAnalyzer.ts
import type {
  ActivityType,
  Difficulty,
  LatLng,
  RouteCandidate,
  RouteMetrics,
  RouteType,
  UserPreferences,
} from "@/types/route";
import { estimateDurationMin, round } from "@/lib/geoUtils";
import type { PathResult, RouteEdge } from "./graph";
import { getEdgeKey } from "./graph";
import {
  bearingDegrees,
  calculateGeometryDistanceM,
  isAfternoonDeparture,
  isNightDeparture,
  normalizePreference,
  resolveTargetDistanceKm,
  wantsClimbing,
  weightedAverage,
} from "./geoUtils";
import { scoreRoute } from "./routeScoring";

export type RouteStrategy = "recommended" | "lowest_elevation" | "exploration" | "park" | "direct";

// Returns the net elevation change for an edge.
// For OSM routes with working elevation API, edges carry real fromAbsElevM/toAbsElevM
// and elevationGainM is their difference. For mock routes and API-failure fallback,
// we estimate using a bearing-based terrain bias (same model as the mock graph).
function estimatedEdgeElevation(edge: RouteEdge): number {
  if (edge.geometry.length < 2) return 0;
  const from = edge.geometry[0];
  const to = edge.geometry[edge.geometry.length - 1];
  const bearing = bearingDegrees(from, to);
  const uphillBias = Math.cos((bearing * Math.PI) / 180) * 0.018 + Math.sin((bearing * Math.PI) / 180) * 0.01;
  return edge.distanceM * uphillBias;
}

function buildElevationData(edges: RouteEdge[]) {
  // Edges with fromAbsElevM/toAbsElevM carry actual Open-Meteo elevation data.
  // Edges with only elevationGainM (non-zero) carry mock-graph bearing estimates.
  // Edges with all zeros have no elevation data — use bearing estimate as fallback.
  const hasAbsoluteElevation = edges.some((e) => e.fromAbsElevM != null);
  const hasAnyElevationData = hasAbsoluteElevation || edges.some((e) => (e.elevationGainM ?? 0) !== 0);

  function netElevChange(edge: RouteEdge): number {
    if (hasAbsoluteElevation) {
      if (edge.fromAbsElevM != null && edge.toAbsElevM != null) {
        return edge.toAbsElevM - edge.fromAbsElevM;
      }
      return edge.elevationGainM ?? 0;
    }
    if (hasAnyElevationData) return edge.elevationGainM ?? 0;
    return estimatedEdgeElevation(edge);
  }

  const startAbsElev = hasAbsoluteElevation
    ? (edges.find((e) => e.fromAbsElevM != null)?.fromAbsElevM ?? 0)
    : 0;

  // Build raw profile, tagging each point as "real" (from API) or estimated.
  // profile[0] is treated as a known anchor when we have any absolute elevation.
  type RawPoint = { distanceKm: number; elevM: number; real: boolean };
  const rawProfile: RawPoint[] = [
    { distanceKm: 0, elevM: Math.round(startAbsElev), real: hasAbsoluteElevation },
  ];
  let accDistM = 0;
  let accElevM = startAbsElev;

  for (const edge of edges) {
    accDistM += edge.distanceM;
    let real: boolean;
    if (hasAbsoluteElevation && edge.toAbsElevM != null) {
      accElevM = edge.toAbsElevM;
      real = true;
    } else {
      accElevM += netElevChange(edge);
      real = false;
    }
    rawProfile.push({ distanceKm: Math.round(accDistM / 10) / 100, elevM: Math.round(accElevM), real });
  }

  // When we have absolute elevation, interpolate linearly between real anchor points so
  // connector edges (which carry no API elevation) don't produce flat plateaus in the chart.
  let profile: Array<{ distanceKm: number; elevM: number }>;
  if (hasAbsoluteElevation) {
    const filled = rawProfile.map((p) => ({ ...p }));
    const realIdx = filled.map((p, i) => (p.real ? i : -1)).filter((i) => i >= 0);

    // Interpolate between consecutive real points.
    for (let r = 0; r + 1 < realIdx.length; r++) {
      const lo = realIdx[r];
      const hi = realIdx[r + 1];
      const fromElev = filled[lo].elevM;
      const toElev = filled[hi].elevM;
      const span = filled[hi].distanceKm - filled[lo].distanceKm;
      for (let j = lo + 1; j < hi; j++) {
        const t = span > 0 ? (filled[j].distanceKm - filled[lo].distanceKm) / span : 0;
        filled[j].elevM = Math.round(fromElev + t * (toElev - fromElev));
      }
    }

    // Hold the last known elevation for any trailing connector points.
    if (realIdx.length > 0) {
      const lastElev = filled[realIdx[realIdx.length - 1]].elevM;
      for (let j = realIdx[realIdx.length - 1] + 1; j < filled.length; j++) {
        filled[j].elevM = lastElev;
      }
    }

    profile = filled.map(({ distanceKm, elevM }) => ({ distanceKm, elevM }));
  } else {
    profile = rawProfile.map(({ distanceKm, elevM }) => ({ distanceKm, elevM }));
  }

  // Compute stats from the interpolated profile: sum positive/negative point-to-point diffs.
  const elevations = profile.map((p) => p.elevM);
  const lowestElevM = Math.min(...elevations);
  const highestElevM = Math.max(...elevations);
  const elevationGainM = Math.round(
    profile.reduce((total, point, i) => total + (i === 0 ? 0 : Math.max(point.elevM - profile[i - 1].elevM, 0)), 0),
  );
  const totalDescentM = Math.round(
    profile.reduce((total, point, i) => total + (i === 0 ? 0 : Math.max(profile[i - 1].elevM - point.elevM, 0)), 0),
  );

  return {
    profile,
    elevationGainM,
    totalDescentM,
    lowestElevM: Math.round(lowestElevM),
    highestElevM: Math.round(highestElevM),
    elevDifferenceM: Math.round(highestElevM - lowestElevM),
  };
}

export interface GeneratedRouteCandidate extends RouteCandidate {
  edgeIds: string[];
  edges: RouteEdge[];
  strategy: RouteStrategy;
}

export interface RouteDraft {
  id: string;
  name: string;
  activity: ActivityType;
  routeType: RouteType;
  path: PathResult;
  waypoints?: LatLng[];
  strategy: RouteStrategy;
  referenceEdgeIds?: Set<string>;
}

function clampScore(value: number) {
  return Math.max(0, Math.min(100, value));
}

function edgeAverage(edges: RouteEdge[], selector: (edge: RouteEdge) => number) {
  return weightedAverage(
    edges.map((edge) => ({
      value: selector(edge),
      weight: Math.max(edge.distanceM, 1),
    })),
    0.5,
  );
}

function scoreDistance(distanceKm: number, targetDistanceKm: number) {
  const missRatio = Math.abs(distanceKm - targetDistanceKm) / Math.max(targetDistanceKm, 0.1);
  return clampScore(100 - missRatio * 160);
}

function scoreElevation(
  activity: ActivityType,
  distanceKm: number,
  elevationGainM: number,
  maxSlopePct: number,
  preferences: UserPreferences,
) {
  const gainPerKm = elevationGainM / Math.max(distanceKm, 0.1);
  const elevationPreference = normalizePreference(preferences.elevationPreference);

  if (wantsClimbing(preferences)) {
    const targetGainPerKm =
      activity === "cycling" ? 14 + elevationPreference * 12 : 22 + elevationPreference * 34;
    const slopePenalty = Math.max(0, maxSlopePct - (activity === "cycling" ? 8 : 14)) * 4;
    return clampScore(100 - Math.abs(gainPerKm - targetGainPerKm) * 1.25 - slopePenalty);
  }

  const slopePenalty = Math.max(0, maxSlopePct - (activity === "cycling" ? 7 : 11)) * 3.5;
  const gainPenalty = gainPerKm * (activity === "cycling" ? 1.2 : 1.05);

  return clampScore(100 - gainPenalty * (0.65 + elevationPreference * 0.75) - slopePenalty);
}

function getReferenceOverlap(edges: RouteEdge[], referenceEdgeIds?: Set<string>) {
  if (!referenceEdgeIds || referenceEdgeIds.size === 0) {
    return 0;
  }

  const sharedDistance = edges.reduce((total, edge) => {
    return referenceEdgeIds.has(getEdgeKey(edge)) ? total + edge.distanceM : total;
  }, 0);
  const totalDistance = edges.reduce((total, edge) => total + edge.distanceM, 0);

  return totalDistance > 0 ? sharedDistance / totalDistance : 0;
}

function getDifficulty(
  activity: ActivityType,
  distanceKm: number,
  elevationGainM: number,
  maxSlopePct: number,
): Difficulty {
  const gainPerKm = elevationGainM / Math.max(distanceKm, 0.1);
  const effort = distanceKm * (activity === "cycling" ? 0.35 : 1) + gainPerKm * 0.18 + maxSlopePct * 0.18;

  if (effort > (activity === "hiking" ? 18 : 13)) {
    return "Hard";
  }

  if (effort > (activity === "cycling" ? 8 : 7)) {
    return "Moderate";
  }

  return "Easy";
}

function buildExplanation(
  route: RouteCandidate,
  preferences: UserPreferences,
  metrics: RouteMetrics,
) {
  const targetDistanceKm = resolveTargetDistanceKm(preferences);
  const reasons = [
    metrics.distanceScore >= 86 ? `closely matches your ${round(targetDistanceKm, 1)} km target` : null,
    metrics.parkScore >= 70 ? "uses park paths and green corridors" : null,
    metrics.elevationScore >= 78 && preferences.routeStyle !== "climbing" ? "keeps elevation controlled" : null,
    metrics.elevationScore >= 78 && preferences.routeStyle === "climbing" ? "adds a measured climbing profile" : null,
    metrics.shadeScore >= 70 ? "prioritizes shade" : null,
    metrics.safetyScore >= 76 ? "avoids higher-stress road segments" : null,
    metrics.explorationScore >= 72 ? "adds legal alternative paths for novelty" : null,
    metrics.sceneryScore >= 74 ? "includes scenic segments" : null,
  ].filter(Boolean);

  const departureNote =
    isAfternoonDeparture(preferences.departureTime) && metrics.shadeScore >= 65
      ? " during your selected afternoon departure time"
      : "";

  if (reasons.length === 0) {
    return `This ${preferences.activity} route balances distance, safety, surface quality, and outdoor appeal.`;
  }

  return `This route was selected because it ${reasons.slice(0, 4).join(", ")}${departureNote}.`;
}

export function analyzeRoute(route: RouteDraft, preferences: UserPreferences): RouteMetrics {
  const edges = route.path.edges;
  const targetDistanceKm = resolveTargetDistanceKm(preferences);
  const distanceKm = round(route.path.distanceM / 1000, 1);
  const elevationGainM = Math.round(edges.reduce((total, edge) => total + Math.max(edge.elevationGainM ?? 0, 0), 0));
  const maxSlopePct = Math.max(0, ...edges.map((edge) => Math.abs(edge.slope ?? 0) * 100));
  const referenceOverlap = getReferenceOverlap(edges, route.referenceEdgeIds);
  const shadePenalty = isAfternoonDeparture(preferences.departureTime) ? 8 : 0;
  const nightSafetyPenalty = isNightDeparture(preferences.departureTime) ? 8 : 0;

  const baseMetrics = {
    distanceScore: scoreDistance(distanceKm, targetDistanceKm),
    elevationScore: scoreElevation(preferences.activity, distanceKm, elevationGainM, maxSlopePct, preferences),
    parkScore: clampScore(edgeAverage(edges, (edge) => edge.parkScore) * 100),
    shadeScore: clampScore(
      edgeAverage(edges, (edge) => edge.shadeScore) * 100 -
        shadePenalty * (1 - edgeAverage(edges, (edge) => edge.shadeScore)),
    ),
    safetyScore: clampScore(
      edgeAverage(edges, (edge) => edge.safetyScore) * 100 -
        nightSafetyPenalty -
        edgeAverage(edges, (edge) => edge.trafficExposure ?? 0.2) * 9,
    ),
    explorationScore: clampScore(
      edgeAverage(edges, (edge) => edge.noveltyScore ?? edge.sceneryScore) * 70 +
        (1 - referenceOverlap) * 30,
    ),
    sceneryScore: clampScore(edgeAverage(edges, (edge) => edge.sceneryScore) * 100),
  };

  return {
    ...baseMetrics,
    totalScore: scoreRoute(
      {
        activity: preferences.activity,
        metrics: {
          ...baseMetrics,
          totalScore: 0,
        },
      },
      preferences,
    ),
  };
}

export function buildRouteCandidate(route: RouteDraft, preferences: UserPreferences): GeneratedRouteCandidate {
  const metrics = analyzeRoute(route, preferences);
  const distanceKm = round(route.path.distanceM / 1000, 1);
  const elevData = buildElevationData(route.path.edges);
  const maxSlopePct = Math.round(Math.max(0, ...route.path.edges.map((edge) => Math.abs(edge.slope ?? 0) * 100)) * 10) / 10;
  const geometryDistanceM = calculateGeometryDistanceM(route.path.geometry);
  const averageSlopePct =
    geometryDistanceM > 0 ? Math.round((elevData.elevationGainM / geometryDistanceM) * 1000) / 10 : undefined;
  const candidate: RouteCandidate = {
    id: route.id,
    source: "generated",
    name: route.name,
    activity: route.activity,
    routeType: route.routeType,
    geometry: route.path.geometry,
    waypoints: route.waypoints,
    distanceKm,
    estimatedDurationMin: estimateDurationMin(route.activity, distanceKm),
    elevationGainM: elevData.elevationGainM,
    totalDescentM: elevData.totalDescentM,
    averageSlopePct,
    maxSlopePct,
    lowestElevM: elevData.lowestElevM,
    highestElevM: elevData.highestElevM,
    elevDifferenceM: elevData.elevDifferenceM,
    elevationProfile: elevData.profile,
    difficulty: getDifficulty(route.activity, distanceKm, elevData.elevationGainM, maxSlopePct),
    metrics,
    explanation: "",
  };

  return {
    ...candidate,
    explanation: buildExplanation(candidate, preferences, metrics),
    edgeIds: route.path.edges.map((edge) => getEdgeKey(edge)),
    edges: route.path.edges,
    strategy: route.strategy,
  };
}

export function routeEdgeIdSet(route: GeneratedRouteCandidate) {
  return new Set(route.edges.map((edge) => getEdgeKey(edge)));
}

File: services/routecraft-api/src/main/java/com/routecraft/api/routing/generator/RouteAnalyzer.java
package com.routecraft.api.routing.generator;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Set;
import java.util.function.ToDoubleFunction;

import com.routecraft.api.routing.graph.GeoUtils;
import com.routecraft.api.routing.graph.RouteEdge;
import com.routecraft.api.routing.model.ActivityType;
import com.routecraft.api.routing.model.Difficulty;
import com.routecraft.api.routing.model.ElevationPoint;
import com.routecraft.api.routing.model.LatLng;
import com.routecraft.api.routing.model.RouteCandidate;
import com.routecraft.api.routing.model.RouteMetrics;
import com.routecraft.api.routing.model.RouteSource;
import com.routecraft.api.routing.model.RouteStyle;
import com.routecraft.api.routing.model.UserPreferences;

public final class RouteAnalyzer {

    private RouteAnalyzer() {
    }

    public static GeneratedRouteCandidate buildRouteCandidate(RouteDraft draft, UserPreferences preferences) {
        RouteMetrics metrics = analyzeRoute(draft, preferences);
        double distanceKm = GeoUtils.round(draft.path().distanceM() / 1000.0, 1);
        ElevationData elev = buildElevationData(draft.path().edges());
        double maxSlopePct = Math.round(maxAbsSlope(draft.path().edges()) * 100 * 10) / 10.0;
        double geometryDistanceM = GeoUtils.calculateGeometryDistanceM(draft.path().geometry());
        Double averageSlopePct = geometryDistanceM > 0
                ? Math.round((elev.elevationGainM / geometryDistanceM) * 1000.0) / 10.0
                : null;
        Difficulty difficulty = getDifficulty(draft.activity(), distanceKm, elev.elevationGainM, maxSlopePct);
        String explanation = buildExplanation(preferences, metrics);

        List<String> edgeIds = new ArrayList<>(draft.path().edges().size());
        for (RouteEdge edge : draft.path().edges()) {
            edgeIds.add(edge.undirectedKey());
        }

        RouteCandidate candidate = new RouteCandidate(
                draft.id(),
                RouteSource.GENERATED,
                draft.name(),
                draft.activity(),
                draft.routeType(),
                draft.path().geometry(),
                draft.waypoints(),
                distanceKm,
                GeoUtils.estimateDurationMin(draft.activity(), distanceKm),
                elev.elevationGainM,
                elev.totalDescentM,
                averageSlopePct,
                maxSlopePct,
                elev.lowestElevM,
                elev.highestElevM,
                elev.elevDifferenceM,
                elev.profile,
                difficulty,
                metrics,
                explanation);

        return new GeneratedRouteCandidate(candidate, edgeIds, draft.path().edges(), draft.strategy());
    }

    public static RouteMetrics analyzeRoute(RouteDraft draft, UserPreferences preferences) {
        List<RouteEdge> edges = draft.path().edges();
        double targetDistanceKm = GeoUtils.resolveTargetDistanceKm(preferences);
        double distanceKm = GeoUtils.round(draft.path().distanceM() / 1000.0, 1);
        double elevationGainM = 0;
        for (RouteEdge edge : edges) {
            elevationGainM += Math.max(edge.elevationGainM() == null ? 0 : edge.elevationGainM(), 0);
        }
        elevationGainM = Math.round(elevationGainM);
        double maxSlopePct = maxAbsSlope(edges) * 100;

        double referenceOverlap = getReferenceOverlap(edges, draft.referenceEdgeIds());
        double shadePenalty = GeoUtils.isAfternoonDeparture(preferences.departureTime()) ? 8 : 0;
        double nightSafetyPenalty = GeoUtils.isNightDeparture(preferences.departureTime()) ? 8 : 0;

        double avgPark = edgeAverage(edges, RouteEdge::parkScore);
        double avgShade = edgeAverage(edges, RouteEdge::shadeScore);
        double avgSafety = edgeAverage(edges, RouteEdge::safetyScore);
        double avgScenery = edgeAverage(edges, RouteEdge::sceneryScore);
        double avgNovelty = edgeAverage(edges, e -> e.noveltyScore() == null ? e.sceneryScore() : e.noveltyScore());
        double avgTraffic = edgeAverage(edges, e -> e.trafficExposure() == null ? 0.2 : e.trafficExposure());

        double distanceScore = scoreDistance(distanceKm, targetDistanceKm);
        double elevationScore = scoreElevation(preferences.activity(), distanceKm, elevationGainM, maxSlopePct, preferences);
        double parkScore = clamp(avgPark * 100, 0, 100);
        double shadeScore = clamp(avgShade * 100 - shadePenalty * (1 - avgShade), 0, 100);
        double safetyScore = clamp(avgSafety * 100 - nightSafetyPenalty - avgTraffic * 9, 0, 100);
        double explorationScore = clamp(avgNovelty * 70 + (1 - referenceOverlap) * 30, 0, 100);
        double sceneryScore = clamp(avgScenery * 100, 0, 100);

        RouteMetrics partial = new RouteMetrics(
                parkScore, shadeScore, safetyScore, explorationScore, sceneryScore,
                elevationScore, distanceScore, 0);
        int totalScore = RouteScoring.scoreRoute(partial, preferences);

        return new RouteMetrics(parkScore, shadeScore, safetyScore, explorationScore, sceneryScore,
                elevationScore, distanceScore, totalScore);
    }

    private static double maxAbsSlope(List<RouteEdge> edges) {
        double max = 0;
        for (RouteEdge edge : edges) {
            double slope = Math.abs(edge.slope() == null ? 0 : edge.slope());
            if (slope > max) {
                max = slope;
            }
        }
        return max;
    }

    private static double edgeAverage(List<RouteEdge> edges, ToDoubleFunction<RouteEdge> selector) {
        List<double[]> entries = new ArrayList<>(edges.size());
        for (RouteEdge edge : edges) {
            entries.add(new double[]{ selector.applyAsDouble(edge), Math.max(edge.distanceM(), 1) });
        }
        return GeoUtils.weightedAverage(entries, 0.5);
    }

    private static double clamp(double value, double min, double max) {
        return Math.max(min, Math.min(max, value));
    }

    private static double scoreDistance(double distanceKm, double targetDistanceKm) {
        double miss = Math.abs(distanceKm - targetDistanceKm) / Math.max(targetDistanceKm, 0.1);
        return clamp(100 - miss * 160, 0, 100);
    }

    private static double scoreElevation(
            ActivityType activity,
            double distanceKm,
            double elevationGainM,
            double maxSlopePct,
            UserPreferences preferences) {
        double gainPerKm = elevationGainM / Math.max(distanceKm, 0.1);
        double elevationPreference = GeoUtils.normalizePreference(preferences.elevationPreference());

        if (GeoUtils.wantsClimbing(preferences)) {
            double targetGainPerKm = activity == ActivityType.CYCLING
                    ? 14 + elevationPreference * 12
                    : 22 + elevationPreference * 34;
            double slopePenalty = Math.max(0, maxSlopePct - (activity == ActivityType.CYCLING ? 8 : 14)) * 4;
            return clamp(100 - Math.abs(gainPerKm - targetGainPerKm) * 1.25 - slopePenalty, 0, 100);
        }

        double slopePenalty = Math.max(0, maxSlopePct - (activity == ActivityType.CYCLING ? 7 : 11)) * 3.5;
        double gainPenalty = gainPerKm * (activity == ActivityType.CYCLING ? 1.2 : 1.05);
        return clamp(100 - gainPenalty * (0.65 + elevationPreference * 0.75) - slopePenalty, 0, 100);
    }

    private static double getReferenceOverlap(List<RouteEdge> edges, Set<String> referenceEdgeIds) {
        if (referenceEdgeIds == null || referenceEdgeIds.isEmpty()) {
            return 0;
        }
        double sharedDistance = 0;
        double totalDistance = 0;
        for (RouteEdge edge : edges) {
            if (referenceEdgeIds.contains(edge.undirectedKey())) {
                sharedDistance += edge.distanceM();
            }
            totalDistance += edge.distanceM();
        }
        return totalDistance > 0 ? sharedDistance / totalDistance : 0;
    }

    private static Difficulty getDifficulty(
            ActivityType activity,
            double distanceKm,
            double elevationGainM,
            double maxSlopePct) {
        double gainPerKm = elevationGainM / Math.max(distanceKm, 0.1);
        double effort = distanceKm * (activity == ActivityType.CYCLING ? 0.35 : 1)
                + gainPerKm * 0.18 + maxSlopePct * 0.18;
        if (effort > (activity == ActivityType.HIKING ? 18 : 13)) {
            return Difficulty.HARD;
        }
        if (effort > (activity == ActivityType.CYCLING ? 8 : 7)) {
            return Difficulty.MODERATE;
        }
        return Difficulty.EASY;
    }

    private static String buildExplanation(UserPreferences preferences, RouteMetrics metrics) {
        List<String> reasons = new ArrayList<>();
        if (metrics.parkScore() >= 70) reasons.add("uses park paths and green corridors");
        if (metrics.elevationScore() >= 78 && preferences.routeStyle() != RouteStyle.CLIMBING) {
            reasons.add("keeps elevation controlled");
        }
        if (metrics.elevationScore() >= 78 && preferences.routeStyle() == RouteStyle.CLIMBING) {
            reasons.add("adds a measured climbing profile");
        }
        if (metrics.shadeScore() >= 70) reasons.add("prioritizes shade");
        if (metrics.safetyScore() >= 76) reasons.add("avoids higher-stress road segments");
        if (metrics.explorationScore() >= 72) reasons.add("adds legal alternative paths for novelty");
        if (metrics.sceneryScore() >= 74) reasons.add("includes scenic segments");

        String departureNote = GeoUtils.isAfternoonDeparture(preferences.departureTime()) && metrics.shadeScore() >= 65
                ? " during your selected afternoon departure time"
                : "";

        if (reasons.isEmpty()) {
            return "Balances distance, safety, surface quality, and outdoor appeal for your "
                    + preferences.activity().value() + ".";
        }

        List<String> top = reasons.subList(0, Math.min(4, reasons.size()));
        String lead = top.get(0);
        lead = Character.toUpperCase(lead.charAt(0)) + lead.substring(1);
        StringBuilder body = new StringBuilder(lead);
        for (int i = 1; i < top.size(); i++) {
            body.append(", ").append(top.get(i));
        }
        return body.toString() + departureNote + ".";
    }

    private static ElevationData buildElevationData(List<RouteEdge> edges) {
        boolean hasAbsoluteElevation = edges.stream().anyMatch(e -> e.fromAbsElevM() != null);
        boolean hasAnyElevationData = hasAbsoluteElevation
                || edges.stream().anyMatch(e -> e.elevationGainM() != null && e.elevationGainM() != 0);

        double startAbsElev = 0;
        if (hasAbsoluteElevation) {
            for (RouteEdge edge : edges) {
                if (edge.fromAbsElevM() != null) {
                    startAbsElev = edge.fromAbsElevM();
                    break;
                }
            }
        }

        List<RawPoint> raw = new ArrayList<>();
        raw.add(new RawPoint(0, Math.round(startAbsElev), hasAbsoluteElevation));
        double accDistM = 0;
        double accElevM = startAbsElev;
        for (RouteEdge edge : edges) {
            accDistM += edge.distanceM();
            boolean real;
            if (hasAbsoluteElevation && edge.toAbsElevM() != null) {
                accElevM = edge.toAbsElevM();
                real = true;
            } else {
                accElevM += netElevChange(edge, hasAbsoluteElevation, hasAnyElevationData);
                real = false;
            }
            raw.add(new RawPoint(Math.round(accDistM / 10) / 100.0, Math.round(accElevM), real));
        }

        // Linear interpolation between real anchors when absolute elevation is available
        if (hasAbsoluteElevation) {
            List<Integer> realIdx = new ArrayList<>();
            for (int i = 0; i < raw.size(); i++) {
                if (raw.get(i).real) realIdx.add(i);
            }
            for (int r = 0; r + 1 < realIdx.size(); r++) {
                int lo = realIdx.get(r);
                int hi = realIdx.get(r + 1);
                double fromElev = raw.get(lo).elevM;
                double toElev = raw.get(hi).elevM;
                double span = raw.get(hi).distanceKm - raw.get(lo).distanceKm;
                for (int j = lo + 1; j < hi; j++) {
                    double t = span > 0 ? (raw.get(j).distanceKm - raw.get(lo).distanceKm) / span : 0;
                    raw.set(j, new RawPoint(raw.get(j).distanceKm,
                            Math.round(fromElev + t * (toElev - fromElev)), false));
                }
            }
            if (!realIdx.isEmpty()) {
                int last = realIdx.get(realIdx.size() - 1);
                double lastElev = raw.get(last).elevM;
                for (int j = last + 1; j < raw.size(); j++) {
                    raw.set(j, new RawPoint(raw.get(j).distanceKm, lastElev, false));
                }
            }
        }

        List<ElevationPoint> profile = new ArrayList<>(raw.size());
        for (RawPoint rp : raw) {
            profile.add(new ElevationPoint(rp.distanceKm, rp.elevM));
        }
        double lowest = profile.stream().mapToDouble(ElevationPoint::elevM).min().orElse(0);
        double highest = profile.stream().mapToDouble(ElevationPoint::elevM).max().orElse(0);
        double gain = 0;
        double descent = 0;
        for (int i = 1; i < profile.size(); i++) {
            double diff = profile.get(i).elevM() - profile.get(i - 1).elevM();
            if (diff > 0) gain += diff;
            else descent -= diff;
        }
        return new ElevationData(profile,
                Math.round(gain),
                Math.round(descent),
                Math.round(lowest),
                Math.round(highest),
                Math.round(highest - lowest));
    }

    private static double netElevChange(RouteEdge edge, boolean hasAbsoluteElevation, boolean hasAnyElevationData) {
        if (hasAbsoluteElevation) {
            if (edge.fromAbsElevM() != null && edge.toAbsElevM() != null) {
                return edge.toAbsElevM() - edge.fromAbsElevM();
            }
            return edge.elevationGainM() == null ? 0 : edge.elevationGainM();
        }
        if (hasAnyElevationData) {
            return edge.elevationGainM() == null ? 0 : edge.elevationGainM();
        }
        return estimatedEdgeElevation(edge);
    }

    private static double estimatedEdgeElevation(RouteEdge edge) {
        if (edge.geometry().size() < 2) {
            return 0;
        }
        LatLng from = edge.geometry().get(0);
        LatLng to = edge.geometry().get(edge.geometry().size() - 1);
        double bearing = GeoUtils.bearingDegrees(from, to);
        double uphillBias = Math.cos((bearing * Math.PI) / 180) * 0.018
                + Math.sin((bearing * Math.PI) / 180) * 0.01;
        return edge.distanceM() * uphillBias;
    }

    private record RawPoint(double distanceKm, double elevM, boolean real) {
    }

    public record ElevationData(
            List<ElevationPoint> profile,
            double elevationGainM,
            double totalDescentM,
            double lowestElevM,
            double highestElevM,
            double elevDifferenceM) {
    }

    public static List<GeneratedRouteCandidate> sortByTotalScoreDescending(List<GeneratedRouteCandidate> routes) {
        List<GeneratedRouteCandidate> sorted = new ArrayList<>(routes);
        sorted.sort(Comparator.comparingDouble((GeneratedRouteCandidate r) -> r.candidate().metrics().totalScore()).reversed());
        return sorted;
    }
}
